graphistry/compute/predicates/numeric.py

```diff
@@ -1,80 +1,162 @@
+
+###
+
+class GT(NumericASTPredicate):
     def __init__(self, val: float) -> None:
         self.val = val
 
     def __call__(self, s: pd.Series) -> pd.Series:
         return s > self.val
 
+    def to_json(self, validate=True) -> dict:
+        if validate:
+            self.validate()
+        return {'type': 'GT', 'val': self.val}
+    
+    @classmethod
+    def from_json(cls, d: dict) -> 'GT':
+        assert 'val' in d
+        out = GT(val=d['val'])
+        out.validate()
+        return out
+
 def gt(val: float) -> GT:
     """
     Return whether a given value is greater than a threshold
     """
     return GT(val)
 
-class LT(ASTPredicate):
+class LT(NumericASTPredicate):
     def __init__(self, val: float) -> None:
         self.val = val
 
     def __call__(self, s: pd.Series) -> pd.Series:
         return s < self.val
 
+    def to_json(self, validate=True) -> dict:
+        if validate:
+            self.validate()
+        return {'type': 'LT', 'val': self.val}
+    
+    @classmethod
+    def from_json(cls, d: dict) -> 'LT':
+        assert 'val' in d
+        out = LT(val=d['val'])
+        out.validate()
+        return out
+
 def lt(val: float) -> LT:
     """
     Return whether a given value is less than a threshold
     """
     return LT(val)
 
-class GE(ASTPredicate):
+class GE(NumericASTPredicate):
     def __init__(self, val: float) -> None:
         self.val = val
 
     def __call__(self, s: pd.Series) -> pd.Series:
         return s >= self.val
 
+    def to_json(self, validate=True) -> dict:
+        if validate:
+            self.validate()
+        return {'type': 'GE', 'val': self.val}
+    
+    @classmethod
+    def from_json(cls, d: dict) -> 'GE':
+        assert 'val' in d
+        out = GE(val=d['val'])
+        out.validate()
+        return out
+
 def ge(val: float) -> GE:
     """
     Return whether a given value is greater than or equal to a threshold
     """
     return GE(val)
 
-class LE(ASTPredicate):
+class LE(NumericASTPredicate):
     def __init__(self, val: float) -> None:
         self.val = val
 
     def __call__(self, s: pd.Series) -> pd.Series:
         return s <= self.val
 
+    def to_json(self, validate=True) -> dict:
+        if validate:
+            self.validate()
+        return {'type': 'LE', 'val': self.val}
+    
+    @classmethod
+    def from_json(cls, d: dict) -> 'LE':
+        assert 'val' in d
+        out = LE(val=d['val'])
+        out.validate()
+        return out
+
 def le(val: float) -> LE:
     """
     Return whether a given value is less than or equal to a threshold
     """
     return LE(val)
 
-class EQ(ASTPredicate):
+class EQ(NumericASTPredicate):
     def __init__(self, val: float) -> None:
         self.val = val
 
     def __call__(self, s: pd.Series) -> pd.Series:
         return s == self.val
 
+    def to_json(self, validate=True) -> dict:
+        if validate:
+            self.validate()
+        return {'type': 'EQ', 'val': self.val}
```

Comment:
There are many calls with this pattern at the end:
```python
return {'type': 'EQ', 'val': self.val}
```
I wonder if we can add some util method in the base class NumericASTPredicate to create that map using self.val internally, so we would have something like this:
```python
return self._to_json('EQ')
```
and the implementation in the base would be:
```python
def _to_json(self, op:string):
  return {'type': op, 'val': self.val}
```

Reply:
ah good idea

like auto-mixin `self.__class__.__name__`

Reply:
Yes it can be a mixing as well.

Reply:
thanks, i was able to cut like 300+ loc this way